feat(consensus): Gossip block parts

greg-szabo opened the pull request:
Part of the work of https://github.com/berachain/cometbft/pull/26.

### Context
The consensus reactor handles communication with peers; for example, sending and receiving votes and block parts.
To support Ethereum EIP-4844 blobs in CometBFT, we need to extend the consensus reactor so it can also transmit and receive blobs alongside the existing consensus data.

### Changes
This PR extends the consensus reactor to support blob handling. Specifically, it introduces a new consensus message, `BlobPartMessage`, which allows the reactor to collect individual blob parts and reconstruct the full blob once all parts are received.
Additionally, it includes a new consensus parameter (originally introduced in #17 and now merged into this PR) that defines the maximum allowed blob size.

greg-szabo committed:
Handle reception of BlobPartMessage

greg-szabo committed:
Gossip blob part to connecting peers (current round only)

greg-szabo committed:
Receive routine fixes

greg-szabo committed:
Clear blob on new round

greg-szabo committed:
add basic test

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -2338,7 +2395,7 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
 		)
 	}
-	if added && cs.ProposalBlockParts.IsComplete() {
+	if added && cs.ProposalBlockParts.IsComplete() && cs.ProposalBlobParts.IsComplete() {
```
Will we ever get here again if we do not have all the blob parts but we do all the block parts? 

That is, say you have all the block parts, you will never get here again, so you will never buil the block?

I would propose to move the check `cs.ProposalBlobParts.IsComplete()` to line 2419 and if the proposal is really complete fire the event, otherwise we get out of this functon and fire it when we have all the blob parts. (Would still have to check this makes sense once looking at other parts of the code).

greg-szabo replied:
Good point, new commit does that.

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -2367,6 +2424,99 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
+		"index", part.Index, "count", count, "total", total, "from", peerID)
+
+	// Todo: Implement blob configuration
+	//maxBytes := cs.state.ConsensusParams.Blob.MaxBytes
```
I am not sure we are adding this to consensus params, it is a hardcoded parameter for now.

greg-szabo replied:
Alessandro has a low priority note at the bottom of the todo list to make it configurable. I kept the Todo here so we know where to go if we want to do it.

I'd like to remove all todos I added before we merge, so we should decide eventually if we want this configurable or not.

alesforz replied:
The max blob size will only change if Ethereum changes it, which I expect will be rare, if it happens at all. That said, I do think it’s a nice-to-have. It’s a simple addition that future-proofs us for that scenario.

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -2367,6 +2424,99 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
+	//		cs.ProposalBlockParts.ByteSize(), maxBytes,
+	//	)
+	//}
+	if added && cs.ProposalBlockParts.IsComplete() && cs.ProposalBlobParts.IsComplete() {
```
Same comment as above for block parts. What if we have all the blob parts but not te block parts, we will essentially never assemble the blob. So here `cs.ProposalBlockParts.IsComplete()` could be moved to 2512. Or changed to a check whether `cs.ProposalBlock != nil`

greg-szabo committed:
different way of checking for complete block

greg-szabo committed:
lint

greg-szabo committed:
lint2

greg-szabo committed:
lint3

jmalicevic commented on internal/consensus/reactor.go:
```diff
@@ -665,6 +668,22 @@ OUTER_LOOP:
+		if part, continueLoop := pickBlobPartToSend(rs, prs, rng); part != nil {
+			// part is not nil: we either succeed in sending it,
+			// or we were instructed not to sleep (busy-waiting)
+			if ps.SendPartSetHasPart(part, prs) || continueLoop {
```
This cannot be reused as `SendPartSetHasPart` marks the block parts as sent.

greg-szabo replied:
Uh, good catch! I'll look into it.

jmalicevic committed:
Properly test whether partset is complete, fixes tests

greg-szabo committed:
Added SendBlobPartSetHasBlobPart

greg-szabo committed:
lint4

greg-szabo committed:
lint5

jmalicevic commented:
Please review `ValidateBasic()` in `part_set.go` , it checks the length against `BlockPartSizeBytes` , it is not correct o check against this for blobs. Maybe we need a flag in the partSet flagging whether this is a blob or a block or pass the parameter against which to check.

greg-szabo committed:
Renamed BlockPartSizeBytes to PartSizeBytes to indicate it is also impacting blobs

greg-szabo commented:
I renamed "BlockPartSizeBytes" to "PartSizeBytes" and added to the comment that it applies for both blocks and blobs. It's not a very elegant solution, but adding a flag would also feel like a hack. I think we could have created a separate "BlobPart" instead of reusing the "Part" for blocks at the beginning but this seems like a good compromise.

greg-szabo committed:
block/blob completeness check fixes

greg-szabo commented:
I reworked the "complete block" checks. I didn't feel comfortable with `nil === IsComplete`. Especially, because if we don't initialize the PartSet somewhere (which apparently we didn't), it would still say `IsComplete`, even if there wasn't even a BlobID. The reworked checks are more in line with Alessandro's documentation.

As it turns out while working with the e2e kvstore, we are not gossiping the blobs. I will have to keep working on this PR.

greg-szabo committed:
trying alternative blob completion check

greg-szabo committed:
blob completion check fixes

greg-szabo committed:
Reverting to IsComplete when nil

jmalicevic commented on types/params.go:
```diff
@@ -19,20 +19,12 @@ const (
-	MaxBlockPartsCount = (MaxBlockSizeBytes / BlockPartSizeBytes) + 1
-
-	// MaxBlobSyzeBytes is the maximum permitted size of a blob.
-	MaxBlobSizeBytes = 800 * 1024 // 800kB
```
This has to stay a separate variable.

jmalicevic replied:
Thinking about it, yes PartSizeBytes should be changed but we need a different maxBlockPartsCount for blobs. Thinking of this, I think we are safer to add a type to the PartSize which we would check (whether its a block or a blob) and then apply the constraints. 

```
type PartSetType int

const (

BlockPartSetT PartSetType = iota
BlobPartSetT 
)
```

```
type PartSet struct {
type PartSetType
}
```

```
func (ps *PartSet) ValidateBasic() {

switch ps.Type {
case BlockPartSetT:

case BlobPartSetT:

default:
Error (or block) 
}
}
```

greg-szabo replied:
Ok, I'll change it.

jmalicevic replied:
WE need `MaxBlobSizeBytes` beacuse it is different from the max block size. The parts can have the same size, that's fine, but the max should be different because it is app specific. And we should use it somewhere (it was defined but not used).

alesforz replied:
While this solution is clean, I’m not convinced it’s necessary. The size of a part (whether it’s a block or a blob) is not configurable. We’ve hard-coded it to 64kB, which works well as the maximum message size.

This is a CometBFT internal implementation detail and is independent of any specific use case. Why assume we need to assign different sizes for blob and block parts? It seems we’re overcomplicating things. Instead, we could simply think in terms of a “part” as a 64kB message, regardless of its type. If you have large objects to send (blocks, blobs, etc.), just split them into 64kB parts. The network and receiver handle all parts the same way, so differentiating them by type is unnecessary. It's the role of the various consensus messages (`*PartMessage` types) to distinguish them, not their size.

That said, the maximum size of an entire object can differ: a block might be 100MB, a blob 800kB, another type 1MB, etc. In these cases, the various `Max*PartsCount` constants will be computed differently, as you say. However, this doesn’t justify adding extra meaning to the `Part` type itself imho.

Does that make sense?

greg-szabo replied:
Ok, understood. I'll remove PartSetType. It will also simplify the ValidateBasic function and get it in line with some interfaces.

jmalicevic commented:
> As it turns out while working with the e2e kvstore, we are not gossiping the blobs. I will have to keep working on this PR. 

The kvstore change should be in as separate PR. But let's not go there until we are sure tests pass here. 

I will see if I can extend `TestStateTimestamp_ProposalMatch` to send a Blob. We should have a basic way of blob testing rather then waiting to have e2e.

jmalicevic committed:
Revert "blob completion check fixes"

This reverts commit de950b1a5a1a24783702fe187f18f094116772fe.

jmalicevic committed:
Revert "trying alternative blob completion check"

This reverts commit 35a2819f669c449b953870e2ad1f1197cb107782.

jmalicevic committed:
Revert "block/blob completeness check fixes"

This reverts commit 9cb911f439c2b2b923a623a242c8ed46811d29ee.

jmalicevic committed:
Revert "Reverting to IsComplete when nil"

This reverts commit dd16f7d1abb089a93b6b02c1ba06334430854392.

jmalicevic committed:
Attempt to fix the checks

jmalicevic committed:
Small fix

greg-szabo committed:
PartSetType for blocks and blobs

greg-szabo committed:
Use partsettype

greg-szabo committed:
supress similar tests as duplicates

jmalicevic committed:
Remove proto decoding for blobs

jmalicevic committed:
Add unit test with blob

jmalicevic committed:
Merge branch 'alesforz/feature/blobs-crgbp' of github.com:alesforz/cometbft into alesforz/feature/blobs-crgbp

jmalicevic committed:
Fix build errors

jmalicevic committed:
Use randStateWithBlob

jmalicevic committed:
Set ProposalBlobPartSetHeader; note this commit also resets them , that is to be revisited

jmalicevic committed:
Unit test to send blob between peers

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -2244,6 +2325,10 @@ func (cs *State) defaultSetProposal(proposal *types.Proposal, recvTime time.Time
 		return ErrProposalTooManyParts
 	}
 
+	if int64(proposal.BlobID.PartSetHeader.Total) > (maxBytes-1)/int64(types.BlockPartSizeBytes)+1 {
```
I could be wrong, but this could be?
```suggestion
	if int64(proposal.BlobID.PartSetHeader.Total) > (maxBytes-1)/int64(types.BlobPartSizeBytes)+1 {
```

greg-szabo committed:
Reintroduced MaxBlobSizeBytes

greg-szabo committed:
lint: removed unused function

greg-szabo committed:
blob existence fixes

greg-szabo committed:
condition simplified, updateHeight clears blob

greg-szabo committed:
Revert "blob existence fixes"

This reverts commit 499371d4be953e0526bd5336f06162f9f4651af3.

greg-szabo committed:
feat(consensus): Configurable max blob size bytes (#17)

* added blob parameters to params.proto

* make proto-gen

* extended ConsensusParams in types/params.go

* test app fixes for configurable blob size

* comment updated in auto-generated protobuf

* unit test fix for blob config

* typo fix

* unit tests for blob bytes config

* add detail to BlobParams doc

* regenerate protobuf files

---------

Co-authored-by: Alessandro Sforzin <alessandro@informal.systems>

alesforz commented on internal/blocksync/msgs.go:
```diff
@@ -14,6 +14,7 @@ const (
 	BlockResponseMessagePrefixSize   = 4
 	BlockResponseMessageFieldKeySize = 1
 	MaxMsgSize                       = types.MaxBlockSizeBytes +
+		types.MaxBlobSizeBytes +
```
this isn't necessary because we never send blobs during blocksync

greg-szabo replied:
Fixed!

alesforz commented on internal/consensus/common_test.go:
```diff
@@ -704,7 +728,7 @@ func ensureRelock(relockCh <-chan cmtpubsub.Message, height int64, round int32)
 }
 
-func ensureProposalWithTimeout(proposalCh <-chan cmtpubsub.Message, height int64, round int32, propID *types.BlockID, timeout time.Duration) {
+func ensureProposalWithTimeout(proposalCh <-chan cmtpubsub.Message, height int64, round int32, propID *types.BlockID, blobID *types.BlobID, timeout time.Duration) {
```
We don't pass around `blobID` as a pointer in the code. Therefore I would do the same in tests.

greg-szabo replied:
Fixed!

alesforz commented on types/part_set.go:
```diff
@@ -175,9 +186,19 @@ func ProtoPartSetHeaderIsZero(ppsh *cmtproto.PartSetHeader) bool {
+
+const (
+	// PartSetTypeBlock indicates the PartSet contains block data.
+	PartSetTypeBlock PartSetType = iota
```
It's best practice to have the first `const` of a enum-like type with `iota+1`. It’s mainly done to avoid accidentally using the zero value as a valid enum. Remember that the default (zero) value for numeric types is 0. If our enum valid values start at 0, then an uninitialized variable of that type will have a valid-looking value, which might lead to bugs.

greg-szabo replied:
The whole thing is removed now, but thanks for spelling out the best practice! It's valuable information.

alesforz commented on types/block.go:
```diff
@@ -281,6 +281,16 @@ func BlockFromProto(bp *cmtproto.Block) (*Block, error) {
 
+// BlobFromProto sets a protobuf Blob to the given pointer.
+// It returns an error if the blob does not exist.
+func BlobFromProto(bp *cmtproto.Blob) (Blob, error) {
```
Is this function necessary? We never serialize a `Blob` to protobuf because it is already a slice of bytes.

greg-szabo replied:
No, removed. Good catch.

greg-szabo committed:
removed PartSetType

greg-szabo committed:
removed unused function, lint

greg-szabo committed:
removed MaxBlobSizeBytes

greg-szabo committed:
No BlobID pointer in tests

greg-szabo committed:
lint

Commit:
Set `ProposalBlob` to empty slice rather than `nil` in `enterNewRound()`

Commit:
Add `TestStateEnterProposeWithBlob()`

Checks that we finish the proposal stage without timing out (i.e., we have the
proposal) for a block with a corresponding blob.

Commit:
Fix incorrect formula computing `MaxBlockPartsCount`.

The previous formula always added one extra part, even when the block size is an
exact multiple of the part size.

Commit:
Rename `BlockPartSizeBytes` to `PartSizeBytes`.

Commit:
Fix `TestNewValidBlockMessageValidateBasic`

Commit:
Add blob in `TestStateFullRound1`

Commit:
Add blob to `TestStateFullRoundNil`

Commit:
Add blob to TestStateFullRound2

Commit:
Fix race condition in test

Commit:
Update util `decideProposal()` with blob.

- Adds a non-empty `BlobID` to proposal
- Returns the blob

Commit:
Add blob to `TestStateLock_NoPOL`

Commit:
Add blob to `TestStateLock_POLUpdateLock`

Commit:
Add blob to `TestStateLock_POLRelock`

Commit:
Add blob to `TestStateLock_PrevoteNilWhenLockedAndMissProposal`

Commit:
Add blob to `TestStateLock_PrevoteNilWhenLockedAndDifferentProposal`

Commit:
Add blob to `TestStateLock_POLDoesNotUnlock`

Commit:
Add blob to `TestStateLock_MissingProposalWhenPOLSeenDoesNotUpdateLock`

Commit:
Add blob to `TestStateLock_MissingProposalWhenPOLForLockedBlock`

Commit:
Add blob to `TestStateLock_DoesNotLockOnOldProposal`

Commit:
Add blob to `TestStateLock_POLSafety1`

Commit:
Add `TestBlob()` to kvstore example application

Commit:
Add blob to `TestStateLock_POLSafety1`

Commit:
Add blob to `TestStateLock_POLSafety2`

Commit:
lintwq

Commit:
Add blob to `TestState_PrevotePOLFromPreviousRound`

Commit:
Add blob to `TestProposeValidBlock`

Commit:
Add blob to `TestSetValidBlockOnDelayedPrevote`

Commit:
`ensureProposalWithTimeout` error string prints BlobID instead of BlockID.

Commit:
Add blob to `TestSetValidBlockOnDelayedProposal`

Commit:
Add blob to `TestResetTimeoutPrecommitUponNewHeight`

Commit:
Add blob to `TestState_MissingProposalValidBlockReceivedTimeout`.

Commit:
Improve blob testing in `TestState_MissingProposalValidBlockReceivedTimeout`

Commit:
Add blob to `TestState_MissingProposalValidBlockReceivedPrecommit`

Commit:
Improve blob testing in `TestStateLock_POLSafety1`

Commit:
Improve blob testing in `TestStateLock_POLSafety2`

Commit:
Add blob to `TestStateTimestamp_ProposalNotMatch`

Commit:
Add blob to `TestStateBadProposal`.

Improve blob testing in `TestStateTimestamp_ProposalNotMatch`

Commit:
Test func `decideProposal` returns app blob instead of generating one.

Commit:
Add blob `nil` checks in several tests

Commit:
Fix `TestResetTimeoutPrecommitUponNewHeight`

Commit:
Add blob to `TestByzantineConflictingProposalsWithPartition`

Commit:
minor refactor and fix incorrect msg
